lib/querystring.js

```diff
@@ -64,70 +64,43 @@ const unhexTable = [
 // a safe fast alternative to decodeURIComponent
 function unescapeBuffer(s, decodeSpaces) {
   var out = Buffer.allocUnsafe(s.length);
-  var state = 0;
-  var n, m, hexchar, c;
-
-  for (var inIndex = 0, outIndex = 0; ; inIndex++) {
-    if (inIndex < s.length) {
-      c = s.charCodeAt(inIndex);
-    } else {
-      if (state > 0) {
-        out[outIndex++] = 37/*%*/;
-        if (state === 2)
-          out[outIndex++] = hexchar;
-      }
-      break;
+  var index = 0;
+  var outIndex = 0;
+  var c, c1, n, m;
+  var maxLength = s.length - 2;
+  // Flag to know if some hex chars have been decoded
+  var hasHex = false;
+  while (index < s.length) {
+    c = s.charCodeAt(index);
+    if (c === 43 /*'+'*/ && decodeSpaces) {
+      out[outIndex++] = 32; // ' '
+      index++;
+      continue;
     }
-    switch (state) {
-      case 0: // Any character
-        switch (c) {
-          case 37: // '%'
-            n = 0;
-            m = 0;
-            state = 1;
-            break;
-          case 43: // '+'
-            if (decodeSpaces)
-              c = 32; // ' '
-            // falls through
-          default:
-            out[outIndex++] = c;
-            break;
-        }
-        break;
-
-      case 1: // First hex digit
-        hexchar = c;
-        n = unhexTable[c];
-        if (!(n >= 0)) {
-          out[outIndex++] = 37/*%*/;
-          out[outIndex++] = c;
-          state = 0;
-          break;
-        }
-        state = 2;
-        break;
-
-      case 2: // Second hex digit
-        state = 0;
-        m = unhexTable[c];
-        if (!(m >= 0)) {
-          out[outIndex++] = 37/*%*/;
-          out[outIndex++] = hexchar;
+    if (c === 37 /*'%'*/ && index < maxLength) {
+      c = s.charCodeAt(++index);
+      n = unhexTable[c];
+      if (n < 0) {
+        out[outIndex++] = 37; // '%'
+      } else {
+        c1 = s.charCodeAt(++index);
+        m = unhexTable[c1];
+        if (m < 0) {
+          out[outIndex++] = 37; // '%'
           out[outIndex++] = c;
-          break;
+          c = c1;
+        } else {
+          hasHex = true;
+          c = n * 16 + m;
         }
-        out[outIndex++] = 16 * n + m;
-        break;
+      }
     }
+    out[outIndex++] = c;
+    index++;
   }
-
-  // TODO support returning arbitrary buffers.
-
-  return out.slice(0, outIndex);
+  return hasHex ? out.slice(0, outIndex) : out;
 }
 
-
```

Comment:
What is the standard? Because the file contains both 2 and 1 line separation between functions.